// appium-base-driver/build/lib/express/static.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.guineaPig = guineaPig;
exports.guineaPigScrollable = guineaPigScrollable;
exports.guineaPigAppBanner = guineaPigAppBanner;
exports.welcome = welcome;
exports.STATIC_DIR = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _path = _interopRequireDefault(require("path"));

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumSupport = require("appium-support");

var _bluebird = _interopRequireDefault(require("bluebird"));

let STATIC_DIR = _path.default.resolve(__dirname, '..', '..', '..', 'static');

exports.STATIC_DIR = STATIC_DIR;

if (_lodash.default.isNull(_path.default.resolve(__dirname).match(/build[/\\]lib[/\\]express$/))) {
  exports.STATIC_DIR = STATIC_DIR = _path.default.resolve(__dirname, '..', '..', 'static');
}

function guineaPigTemplate(_x, _x2, _x3) {
  return _guineaPigTemplate.apply(this, arguments);
}

function _guineaPigTemplate() {
  _guineaPigTemplate = (0, _asyncToGenerator2.default)(function* (req, res, page) {
    let delay = parseInt(req.params.delay || req.query.delay || 0, 10);
    let throwError = req.params.throwError || req.query.throwError || '';
    let params = {
      throwError,
      serverTime: parseInt(Date.now() / 1000, 10),
      userAgent: req.headers['user-agent'],
      comment: 'None'
    };

    if (req.method === 'POST') {
      params.comment = req.body.comments || params.comment;
    }

    _logger.default.debug(`Sending guinea pig response with params: ${JSON.stringify(params)}`);

    if (delay) {
      _logger.default.debug(`Waiting ${delay}ms before responding`);

      yield _bluebird.default.delay(delay);
    }

    res.set('content-type', 'text/html');
    res.cookie('guineacookie1', 'i am a cookie value', {
      path: '/'
    });
    res.cookie('guineacookie2', 'cookié2', {
      path: '/'
    });
    res.cookie('guineacookie3', 'cant access this', {
      domain: '.blargimarg.com',
      path: '/'
    });
    res.send((yield getTemplate(page))(params));
  });
  return _guineaPigTemplate.apply(this, arguments);
}

function guineaPig(_x4, _x5) {
  return _guineaPig.apply(this, arguments);
}

function _guineaPig() {
  _guineaPig = (0, _asyncToGenerator2.default)(function* (req, res) {
    return yield guineaPigTemplate(req, res, 'guinea-pig.html');
  });
  return _guineaPig.apply(this, arguments);
}

function guineaPigScrollable(_x6, _x7) {
  return _guineaPigScrollable.apply(this, arguments);
}

function _guineaPigScrollable() {
  _guineaPigScrollable = (0, _asyncToGenerator2.default)(function* (req, res) {
    return yield guineaPigTemplate(req, res, 'guinea-pig-scrollable.html');
  });
  return _guineaPigScrollable.apply(this, arguments);
}

function guineaPigAppBanner(_x8, _x9) {
  return _guineaPigAppBanner.apply(this, arguments);
}

function _guineaPigAppBanner() {
  _guineaPigAppBanner = (0, _asyncToGenerator2.default)(function* (req, res) {
    return yield guineaPigTemplate(req, res, 'guinea-pig-app-banner.html');
  });
  return _guineaPigAppBanner.apply(this, arguments);
}

function welcome(_x10, _x11) {
  return _welcome.apply(this, arguments);
}

function _welcome() {
  _welcome = (0, _asyncToGenerator2.default)(function* (req, res) {
    let params = {
      message: 'Let\'s browse!'
    };

    _logger.default.debug(`Sending welcome response with params: ${JSON.stringify(params)}`);

    res.send((yield getTemplate('welcome.html'))(params));
  });
  return _welcome.apply(this, arguments);
}

function getTemplate(_x12) {
  return _getTemplate.apply(this, arguments);
}

function _getTemplate() {
  _getTemplate = (0, _asyncToGenerator2.default)(function* (templateName) {
    let content = yield _appiumSupport.fs.readFile(_path.default.resolve(STATIC_DIR, 'test', templateName));
    return _lodash.default.template(content.toString());
  });
  return _getTemplate.apply(this, arguments);
}require('source-map-support').install();
