// appium-android-driver/build/index.js
#!/usr/bin/env node
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CHROMIUM_WIN = exports.WEBVIEW_BASE = exports.WEBVIEW_WIN = exports.NATIVE_WIN = exports.webviewHelpers = exports.commonCapConstraints = exports.startServer = exports.AndroidDriver = exports.androidCommands = exports.androidHelpers = exports.default = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _yargs = _interopRequireDefault(require("yargs"));

var _asyncbox = require("asyncbox");

var server = _interopRequireWildcard(require("./lib/server"));

var driver = _interopRequireWildcard(require("./lib/driver"));

var androidHelperIndex = _interopRequireWildcard(require("./lib/android-helpers"));

var commandIndex = _interopRequireWildcard(require("./lib/commands/index"));

var webview = _interopRequireWildcard(require("./lib/webview-helpers"));

var caps = _interopRequireWildcard(require("./lib/desired-caps"));

const startServer = server.startServer;
exports.startServer = startServer;
const DEFAULT_HOST = "localhost";
const DEFAULT_PORT = 4723;

function main() {
  return _main.apply(this, arguments);
}

function _main() {
  _main = (0, _asyncToGenerator2.default)(function* () {
    let port = _yargs.default.argv.port || DEFAULT_PORT;
    let host = _yargs.default.argv.host || DEFAULT_HOST;
    return yield startServer(port, host);
  });
  return _main.apply(this, arguments);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

const AndroidDriver = driver.AndroidDriver;
exports.AndroidDriver = AndroidDriver;
const webviewHelpers = webview.helpers,
      NATIVE_WIN = webview.NATIVE_WIN,
      WEBVIEW_WIN = webview.WEBVIEW_WIN,
      WEBVIEW_BASE = webview.WEBVIEW_BASE,
      CHROMIUM_WIN = webview.CHROMIUM_WIN;
exports.CHROMIUM_WIN = CHROMIUM_WIN;
exports.WEBVIEW_BASE = WEBVIEW_BASE;
exports.WEBVIEW_WIN = WEBVIEW_WIN;
exports.NATIVE_WIN = NATIVE_WIN;
exports.webviewHelpers = webviewHelpers;
const commonCapConstraints = caps.commonCapConstraints;
exports.commonCapConstraints = commonCapConstraints;
const androidCommands = commandIndex.commands;
exports.androidCommands = androidCommands;
const androidHelpers = androidHelperIndex.helpers;
exports.androidHelpers = androidHelpers;
var _default = AndroidDriver;
exports.default = _default;require('source-map-support').install();
