// appium-android-driver/build/lib/uiautomator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UiAutomator = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _events = _interopRequireDefault(require("events"));

var _appiumSupport = require("appium-support");

const log = _appiumSupport.logger.getLogger('UiAutomator');

class UiAutomator extends _events.default.EventEmitter {
  constructor(adb) {
    if (!adb) {
      log.errorAndThrow('adb is required to instantiate UiAutomator');
    }

    super();
    this.adb = adb;
    this.tempPath = '/data/local/tmp/';
  }

  start(uiAutomatorBinaryPath, className, startDetector, ...extraParams) {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let processIsAlive;

      try {
        log.debug('Starting UiAutomator');

        _this.changeState(UiAutomator.STATE_STARTING);

        log.debug('Parsing uiautomator jar');

        let jarName = _this.parseJarNameFromPath(uiAutomatorBinaryPath);

        yield _this.adb.push(uiAutomatorBinaryPath, _this.tempPath);
        yield _this.killUiAutomatorOnDevice();
        log.debug('Starting UIAutomator');
        let args = ['shell', 'uiautomator', 'runtest', jarName, '-c', className, ...extraParams];
        _this.proc = _this.adb.createSubProcess(args);

        _this.proc.on('exit', (code, signal) => {
          processIsAlive = false;

          if (_this.state !== UiAutomator.STATE_STOPPED && _this.state !== UiAutomator.STATE_STOPPING) {
            let msg = `UiAutomator exited unexpectedly with code ${code}, ` + `signal ${signal}`;
            log.error(msg);
          } else if (_this.state === UiAutomator.STATE_STOPPING) {
            log.debug('UiAutomator shut down normally');
          }

          _this.changeState(UiAutomator.STATE_STOPPED);
        });

        yield _this.proc.start(startDetector);
        processIsAlive = true;

        _this.changeState(UiAutomator.STATE_ONLINE);

        return _this.proc;
      } catch (e) {
        _this.emit(UiAutomator.EVENT_ERROR, e);

        if (processIsAlive) {
          yield _this.killUiAutomatorOnDevice();
          yield _this.proc.stop();
        }

        log.errorAndThrow(e);
      }
    })();
  }

  shutdown() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      log.debug('Shutting down UiAutomator');

      if (_this2.state !== UiAutomator.STATE_STOPPED) {
        _this2.changeState(UiAutomator.STATE_STOPPING);

        yield _this2.proc.stop();
      }

      yield _this2.killUiAutomatorOnDevice();

      _this2.changeState(UiAutomator.STATE_STOPPED);
    })();
  }

  parseJarNameFromPath(binaryPath) {
    let reTest = /.*(\/|\\)(.*\.jar)/.exec(binaryPath);

    if (!reTest) {
      throw new Error(`Unable to parse jar name from ${binaryPath}`);
    }

    let jarName = reTest[2];
    log.debug(`Found jar name: '${jarName}'`);
    return jarName;
  }

  changeState(state) {
    log.debug(`Moving to state '${state}'`);
    this.state = state;
    this.emit(UiAutomator.EVENT_CHANGED, {
      state
    });
  }

  killUiAutomatorOnDevice() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        yield _this3.adb.killProcessesByName('uiautomator');
      } catch (e) {
        log.warn(`Error while killing uiAutomator: ${e}`);
      }
    })();
  }

}

exports.UiAutomator = UiAutomator;
UiAutomator.EVENT_ERROR = 'uiautomator_error';
UiAutomator.EVENT_CHANGED = 'stateChanged';
UiAutomator.STATE_STOPPING = 'stopping';
UiAutomator.STATE_STOPPED = 'stopped';
UiAutomator.STATE_STARTING = 'starting';
UiAutomator.STATE_ONLINE = 'online';
var _default = UiAutomator;
exports.default = _default;require('source-map-support').install();
